### PR TITLE
refactor: delete kakarotethapi and kakarotstarknetapi, refactor into …

### PR DESCRIPTION
**Pull request by Eikix**
Actions:
- delete `KakarotEthApi` and `KakarotStarknetApi` trait. They're adding complexity when no other struct than `KakarotClient` will implement this trait.
- group all functions inside the implementation of KakarotClient
- stop using dynamic dispatch (`&dyn KakarotEthApi`) where `KakakotEthApi` was used, replace by a long-lived reference (`&KakarotClient<P>`, where P is 'static, or an `Arc<KakarotClient<P>>`
- inline some functions of the implementation of the `kakarotclient` inside the eth-rpc, as they're only used once and we don't need this level of abstraction
- delete useless tests
- add an e2e test in eth-rpc
- delete assert helpers and unused serde implementation

Time spent on this PR: 2 days

Resolves: #642 #651 #648

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/core/src/client/tests/mod.rs [hunk 1]
```diff
@@ -194,105 +109,3 @@ async fn test_gas_price() {
...
-
-    // When
-    let logs = client.get_logs(filter).await.unwrap();
-
-    // Then
-    assert_eq!(logs, FilterChanges::Empty);
-}
-
-#[tokio::test]
-async fn test_get_logs() {
```
Comment (Eikix) [inline crates/core/src/client/tests/mod.rs:239]: A bit unlucky that we loose this test, but I'm thinking of:
- For integration: use hive
- For e2e, we can use Katana and go through all the steps in a tx -> run a deploy tx for a counter sc, check logs, call the count() method, increment the counter, call the count() method, assert

Comment (Eikix) [reply crates/core/src/client/tests/mod.rs:239]: Replaced in eth-rpc -> e2e test for get_logs

File: crates/eth-rpc/tests/integration.rs [hunk 2]
```diff
@@ -1,31 +1,136 @@
...
+    // ⚠️ Only one test with a Katana fixture can run at a time.
+    // When trying to run two tests with a Katana fixture, the second test will fail with:
+    // `thread 'test_erc20' panicked at 'Failed to start the server: Os { code: 98, kind:
+    // AddrInUse, message: "Address already in use" }'`
     #[rstest]
     #[awt]
     #[tokio::test(flavor = "multi_thread")]
-    async fn test_get_block_rpc(#[future] katana: Katana) {
-        // Start the Kakarot RPC server
+    async fn test_erc20(#[future] katana: Katana) {
```
Comment (greged93) [inline crates/eth-rpc/tests/integration.rs:32]: I find that this function contains a lot of asserts, could we perhaps have the expected return inside a json file and just deserialize the content and compare?

**Review by greged93 — COMMENTED** [review]
couple of comments, nice cleaning ! Also CI is failing :(

File: crates/eth-rpc/tests/integration.rs [hunk 3]
```diff
@@ -1,31 +1,136 @@
...
     use kakarot_test_utils::sequencer::Katana;
     use reth_primitives::U64;
     use rstest::*;
 
+    abigen!(ERC20, "tests/contracts/ERC20/IERC20.json");
+
+    // ⚠️ Only one test with a Katana fixture can run at a time.
+    // When trying to run two tests with a Katana fixture, the second test will fail with:
+    // `thread 'test_erc20' panicked at 'Failed to start the server: Os { code: 98, kind:
+    // AddrInUse, message: "Address already in use" }'`
```
[the remark this answers is not in the document]
Comment (Eikix) [reply crates/eth-rpc/tests/integration.rs:28]: Ah okay, then I'm not sure, it only fails here when there are two tests that run

File: crates/core/src/client/tests/mod.rs [hunk 4]
```diff
@@ -1,34 +1,17 @@
 use std::str::FromStr;
```
[the remark this answers is not in the document]
Comment (Eikix) [reply crates/core/src/client/tests/mod.rs:1]: I deleted tests for methods I removed from KakarotClient

File: crates/core/tests/api.rs [hunk 5]
```diff
@@ -118,95 +113,4 @@ mod tests {
             .unwrap();
         assert_eq!(U256::from(1), count);
     }
-
-    #[rstest]
```
[the remark this answers is not in the document]
Comment (Eikix) [reply crates/core/tests/api.rs:122]: I removed get_logs so I removed the corresponding test as well

File: crates/eth-rpc/tests/integration.rs (hunk 2 above)
Comment (Eikix) [reply crates/eth-rpc/tests/integration.rs:32]: what is the advantage of such an approach?

File: crates/eth-rpc/tests/integration.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Eikix) [reply crates/eth-rpc/tests/integration.rs:28]: okay makes sense will modify the comment

**Review by greged93 — APPROVED** [review]

**Pull request merged**